### PR TITLE
Support more complex view patterns in pointwise ops.

### PR DESCRIPTION
**Pull request by csarofeen**
This PR is fairly complex, will try to describe some of the things going on in it here:

1) Adds `ComputeAtMap::areExactExprs(Expr* expr_1, Expr* expr_2)` which checks if inputs and outputs of a transformation are in exact map or are matching (like the factor of a split). Order of inputs/outputs is important.

2) Adds `void ComputeAtMap::buildUniqueExactExprMaps()` which calculates `definitions` and `uses` on the exact map. This is computed by grabbing all the definitions of every `IterDomain` in every exact disjoint set (definitions entry we construct is per exact disjoint set), and deduplicating them (within each disjoint sets entry) based on `areExactExprs` to leave a unique series of expression describing all possible definitions constructing any `IterDomain` in the exact disjoint set. Uses are constructed directly from the exact unique definitions.

3) Adds `std::vector<Expr*> uniqueExactDefinitions(IterDomain* id)` and `std::vector<Expr*> uniqueExactUses(IterDomain* id)` which just returns the above entries we make.

The above functions are the basis for `IterDomainGraph` traversals we will start using for analyses. 

4) Disables persistent buffer projection when there's view ops in a normalization

5) Adds `registry.cpp::bool requiresForwardViewReplay(Fusion* fusion, ComputeAtMap& ca_map)` which checks if we can successfully propagate all view transformations in the DAG. This will return true if we can take the output of all view expressions (one by one), and propagate their view transformations through the entire DAG without causing a conflict. A conflict in this is if we propagate forward a series of transformations across a view that doesn't match the view transformations. This type of propagation would require "reversible views". 

This pass is implemented by checking every `rfactor` domain in the compute at map, and making sure they're all only used in a single transformation. In other words if we have two views consuming the same input, we check that they're consumed by exact matching expressions. Since we can always back propagate through views, then we don't have to worry about definitions producing any rfactor domains are exact.

6) Adds 
```.cpp
registry.cpp::bool reductionInterferingView(
    Fusion* fusion,
    const ComputeAtMap& ca_map,
    TensorView* reduction_reference)
```
which looks at the reference provided and the compute at map to make sure none of the view operations would be across IterDomains that we would schedule as independent in the normalization scheduler. In other words if we have a 3D domain `I0, R1, I2` we're detecting there's no operations that would merge any domains derived from these 3 domains. If we have `R0, R1, I2, R3` then it's fine to have view operations that merge domains derived from `R0` and `R1` but not (for example) across `R1` and `I2` or even `R1` and `R3`.

7) Enables view scheduling for normalization and reduction fusions as long as the check (5) and (6) pass.

8) Implemented `scheduler_utilss::propagateViewTransforms(Fusion* fusion, const ComputeAtMap& ca_map)` which requires the properties enforced by (5) and (6) on the fusion. This will go through every view one by one and will propagate the view transformations through the entire DAG.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp [hunk 1]
```diff
@@ -194,7 +202,37 @@ void IterDomainGraph::mapThroughExpr(Expr* first, Expr* second, bool forward) {
 
 namespace {
 
-// Returns a pair of mapped IDs
+// Returns the first pair of id's in ids detected to match eachother on the
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/compute_at_map.cpp:205]: Just added some notes for thoughts on how we should work on this.

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp [hunk 2]
```diff
@@ -1396,26 +1396,6 @@ std::vector<TensorView*> getInputsOutputsWithInnerDim(
   return vectorizable_tensors;
 }
 
-namespace {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.cpp:1399]: Just moved this to make it accessible.

**Comment by naoyam** [comment]
Is `ComputeAtMap::areExactExprs` effectively the same as using `BestEffortReplay` and checking if everything is replayed without forwarding?

**Comment by naoyam** [comment]
> Is `ComputeAtMap::areExactExprs` effectively the same as using `BestEffortReplay` and checking if everything is replayed without forwarding?

Oh, this is for `IterDomain` expressions. Why not creating something more generic that can work with any expression types? 😉

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp [hunk 3]
```diff
@@ -1052,6 +1091,124 @@ void ComputeAtMap::buildConcreteIds() {
...
+    if (expr_1->outputs()[output_i]->isA<IterDomain>() &&
+        !areMapped(
+            expr_1->outputs()[output_i]->as<IterDomain>(),
+            expr_2->outputs()[output_i]->as<IterDomain>(),
+            IdMappingMode::EXACT)) {
+      // Outputs don't exact map in the right order
+      return false;
+    }
+  }
+  // Expr's are almost exact mapped transforms
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: nit: what do you mean by "almost"? Aren't they perfectly mapped?

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp [hunk 4]
```diff
@@ -1052,6 +1091,124 @@ void ComputeAtMap::buildConcreteIds() {
...
+
+void ComputeAtMap::buildUniqueExactExprMaps() {
+  // Start by building definitions
+  for (const auto& disjoint_set_shared_ptr :
+       id_graph_.exactNodes().disjointSets()) {
+    std::vector<Expr*> definitions;
+
+    // N^2 in number of unique transformations, this might be better to do
+    // when generating the map.
+    IterDomain* concrete_id = nullptr;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: This doesn't seem to be used

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp [hunk 5]
```diff
@@ -1052,6 +1091,124 @@ void ComputeAtMap::buildConcreteIds() {
...
+            IdMappingMode::EXACT)) {
+      // Outputs don't exact map in the right order
+      return false;
+    }
+  }
+  // Expr's are almost exact mapped transforms
+  return true;
+}
+
+void ComputeAtMap::buildUniqueExactExprMaps() {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at_map.cpp:1138]: Not a blocker, but since this is such an important analysis, shouldn't we have some unit tests? I wonder what would be interesting tests to exercise corner cases.

**Review by naoyam — COMMENTED** [review]
Just on compute_at_map for now

File: torch/csrc/jit/codegen/cuda/scheduler/utils.h [hunk 6]
```diff
@@ -289,6 +290,41 @@ std::vector<TensorView*> getInputsOutputsWithInnerDim(
     bool inner_only,
     bool vectorize_pass);
 
+// Holder return struct for the below function.
+struct DisjointViewSetInfo {
+  // const* to the disjoint set in disjoint_view_set passed in to
+  // getDisjointViewSetsOf each iterdomain in the rfactor of ref is mapped to.
+  //
+  // WARNING: these pointers are relative to the disjoint_view_set reference
+  // passed into getDisjointViewSetsOf it's the user's responsibillity to
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.h]: typo: responsibility

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 7]
```diff
@@ -1004,12 +1250,12 @@ class ReductionScheduler : public SchedulerEntry {
       return false;
     }
 
-    if (findTransposeOps(fusion).size() > 0) {
-      // Use pointwise logic
-      scheduler_debug_utils::canScheduleRejectReason(
-          ScheduleHeuristic::Reduction, "No support for transpose op");
-      return false;
-    }
+    // if (findTransposeOps(fusion).size() > 0) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: Can this part just be removed?

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 8]
```diff
@@ -446,6 +446,261 @@ bool isConnectedFusionGraph(Fusion* fusion) {
...
+          std::all_of(
+              def_outs.begin(),
+              def_outs.end(),
+              [](IterDomain* id) { return id->isRFactorProduct(); }),
+          "This function does not support outputs of transformations with mismatching rfactor flags. ",
+          "If one output is rfactor all should be rfactor.");
+
+      // There could be a transformation where the inputs are rfactor
+      // dimensions, but outputs are not, just ignore those expressions as
+      // they're scheduling based after of the view definitions.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: Unclear what this means. Looks like the code below actually ignores transformations that have non-rfactor inputs.

When does an expression generate an rfactor output with non-rfactor inputs?

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 9]
```diff
@@ -446,6 +446,261 @@ bool isConnectedFusionGraph(Fusion* fusion) {
...
+  // tv1 root: [I0, I1rf, I2rf] -> rfactor [I0, I1*I2rf]
+  //
+  // Here we can see I1rf is used in two view transformations, one to I0*I1rf,
+  // and the other to I1*I2rf.
+
+  // Track the transformation each exact disjoint rfactor set is used in. If
+  // more than one is detected we can't support transforming the fusion into a
+  // consistent format.
+  std::unordered_map<std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>, Expr*>
+      unique_exact_uses_;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: nit: having a trailing underscore even though this is just a local variable is a little confusing

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 10]
```diff
@@ -446,6 +446,261 @@ bool isConnectedFusionGraph(Fusion* fusion) {
...
+        continue;
+      }
+
+      // Check which definition in the unique exact definition set this
+      // definition matches to:
+      for (auto unique_def : unique_defs) {
+        if (ca_map.areExactExprs(rfactor_def, unique_def)) {
+          // Check if we already have an expression that consumes an
+          // equivalent of any of the input rfactor domains. If so and it's
+          // not the already registered transformation, return false
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: return true?

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 11]
```diff
@@ -446,6 +446,261 @@ bool isConnectedFusionGraph(Fusion* fusion) {
...
+      auto def_inps = ir_utils::filterByType<IterDomain>(rfactor_def->inputs());
+      if (!std::all_of(def_inps.begin(), def_inps.end(), [](IterDomain* id) {
+            return id->isRFactorProduct();
+          })) {
+        continue;
+      }
+
+      // Check which definition in the unique exact definition set this
+      // definition matches to:
+      for (auto unique_def : unique_defs) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:549]: Why does this loop over all the unique defs? By definition, there's only one def that is `areExactExprs` with `rfactor_def`, right? If so, can the use of `unique_def` at lines 564 and 566 be replaced with `rfactor_def`?

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 12]
```diff
@@ -446,6 +446,261 @@ bool isConnectedFusionGraph(Fusion* fusion) {
...
+          break;
+        }
+      }
+    }
+  }
+  // No inconsistent rfactor uses found, we can safely transform this graph.
+  return false;
+}
+
+// Returns if view intefers with how we want to treat the reference, being at
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: typo: interferes

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 13]
```diff
@@ -1336,23 +1585,32 @@ class PersistentKernelScheduler : public SchedulerEntry {
...
+      // that changes, this needs to be changed.
+      if (reductionInterferingView(fusion, ca_map, reduction_tvs[0])) {
+        scheduler_debug_utils::canScheduleRejectReason(
+            ScheduleHeuristic::Persistent,
+            "View may interfere with normalization scheduling.");
+        return false;
+      }
     }
 
+    // if (findTransposeOps(fusion).size() > 0) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: Same here

File: torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp [hunk 14]
```diff
@@ -1724,6 +1724,100 @@ TEST_F(NVFuserTest, FusionViewMagicSchedule5_CUDA) {
   testValidate(&fusion, cg_outputs, {t0, t3}, {t6}, __LINE__, __FILE__);
 }
 
+// View with 3D reduction scheduling
+TEST_F(NVFuserTest, FusionViewMagicSchedule6_CUDA) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp:1726]: 🎉

File: torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp [hunk 15]
```diff
@@ -1724,6 +1724,100 @@ TEST_F(NVFuserTest, FusionViewMagicSchedule5_CUDA) {
...
+  auto t1 = sin(t0);
+  auto t2 = at::native::view(t1, {v * w, x, y * z});
+  at::Tensor t3 = at::randn({v, w, x, z, y}, options);
+  auto t4 = cos(t3);
+  auto t5 = at::native::view(t4, {v * w, x, y * z});
+  auto t7 = add(t2, t5).sum(2).sum(0);
+
+  FusionExecutorCache executor_cache(std::move(fusion_ptr));
+  // Collect the heuristic params
+  executor_cache.profile(true);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp:1759]: Why is this necessary?

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: I think the first time I did this it may have been on the almost exact map instead of exact map, I don't know. Removing comment as it should be clear what true means here at this point.

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp:1138]: Yes, but at this point I rewrote the concrete ID pass using it as well as other heavy weight analysis. It's not explicitly tested, but at this point it seems fairly well implicitly tested. As more systems gravitate towards it I think it will get even better test coverage.

I also don't really know what we'd want to test.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp (hunk 8 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: Yeah, this should be an assert. There shouldn't be a mechanism to introduce transformations producing an rfactor domain from a non-rfactor domain. At least none I can think of.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp (hunk 11 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:549]: I'm hesitant to make that change. Although that should be true for any non-scheduled fusion group, I think eventually we would like to do model parallel scheduling before the scheduling stage. There's a lot that would have to be done to support this, but this incantation shouldn't hurt and would support such scenarios.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp (hunk 15 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp:1759]: copy and past, clearing

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp:1759]: oh no, it's needed for:
```
  TORCH_CHECK(!executor_cache.getMostRecentKernelRuntime()->isSegmented());
  TORCH_CHECK(executor_cache.getMostRecentExecutorInfo()
                  .params->isA<ReductionParams>());
```

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp:1138]: Yeah, agreed. The existing tests should be probably sufficient unless there's any specific corner case that may not be exercised by the tests.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp (hunk 11 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:549]: OK, makes sense. In that case, if model parallel scheduling would rfactor reductions, I wonder if any of the view scheduling code would need to change as anything between root domains and rfactor domains are considered view transformations, which is not true if reduction rfactor is done. Reduction rfactor transformations must also be retained, so they have the same constraints as views, so maybe that's actually just fine?

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for updating the PR.
